### PR TITLE
Improve charmap structure with trie.

### PR DESCRIPTION
**Pull request by jidoc01**
The previous structure of charmap used brute-force string-comparison to convert the strings from its user. It always compares the given string to all of the strings in charmap, which is very costly in a huge project.

For its improvement, I changed its structure into trie, which is being used in many string-processing areas. And thanks to the comments from kind people, I could edit the flaws in it, and committed it again now.

Sorry to bother you! I'm new to this github thing :p

### CODE REVIEW

File: include/asm/charmap.h [hunk 1]
```diff
@@ -11,16 +11,27 @@
 
 #include <stdint.h>
 
-#define MAXCHARMAPS	512
+#define MAXCHARMAPS		512
```
Comment (dbrotz) [inline include/asm/charmap.h]: This extra tab has caused the columns to be misaligned.

File: src/asm/charmap.c [hunk 2]
```diff
@@ -55,91 +54,104 @@ int32_t charmap_Add(char *input, uint8_t output)
...
-		charmap->output[charmap->count] = output;
 	}
-	return ++charmap->count;
+
+	/* prevent duplicated keys by accepting only first key-value pair.  */
+	if (curr_node->isCode)
+		return charmap->charCount;
+
+	curr_node->code 	= output;
+	curr_node->isCode 	= 1;
```
Comment (dbrotz) [inline src/asm/charmap.c]: This alignment of `=` signs isn't consistent with the rest of the code.

File: src/asm/charmap.c [hunk 3]
```diff
@@ -55,91 +54,104 @@ int32_t charmap_Add(char *input, uint8_t output)
...
+	struct Charmap 	*charmap;
+	struct Charnode	*charnode;
+
+	char *output;
+	char outchar[8];
 
-	char outchar[CHARMAPLENGTH + 1];
-	char *buffer;
 	int32_t i, j, length;
+	uint8_t v, w;
```
Comment (dbrotz) [inline src/asm/charmap.c]: The single letter variable names are a bit hard to follow. `i` is OK, since it's just a loop counter, but the `j`, `v`, and `w` variables could have better names. This is just a suggestion.

File: src/asm/charmap.c [hunk 4]
```diff
@@ -55,91 +54,104 @@ int32_t charmap_Add(char *input, uint8_t output)
...
+
 		*input += j;
 	}
-	*input = buffer;
+
+	*input = output;
+
 	return length;
-}
+}
```
Comment (dbrotz) [inline src/asm/charmap.c]: You removed the newline at the end of the file.

**Comment by jidoc01** [comment]
Thanks for reviewing. I edited the things you mentioned, and pushed again.
And I'm uploading a simple benchmark here.

I tested the compilation time of pokecrystal on my common labtop. I used laptop for showing the differences efficiently.

| Time(sec) | Prev. Version | Trie Version |
| ---------- | ---------- | ---------- |
| audio.asm | 2.83 | 2.24 |
| home.asm | 0.20 | 0.26 |
| main.asm | 3.83 | 3.46 |
| common.asm | 0.29 | 0.23 |
| map_data.asm | 2.27 | 2.27 |

Its notable points are below:
1. It shows improvement in all the cases, except the second one.
2. The difference in the second case came from the time for charmap to initialize itself once, which means that the difference is very tiny in overall benefits in a project.
3. There is at most about 0.6 sec improvement on each file, and it leads to a big one in the entire files.

It's just a example for its efficiency. I think there could be an 'overhead' in relatively very small projects, for their small amount to process, and its initializing time. But I expect much improvement in many projects with large-scaled source files.

Thanks.

**Comment by jidoc01** [comment]
Thanks. Glad to hear that :)

And, it's a different matter, I've found a critical bug in processing of comments.
This bug comes out when you comment something with the character ';', and include the quotation mark in it. For example,

ld a, 3
push hl
; This is a comment.
; and I will show you an "error"!

In the case above, you can see an unknown error as a syntax error, but actually it's not a syntax error. The latest version of rgbds compiler has a step to parse the given source to convert its line endings to a unified one, and it processes quotation marks even before it processes the comments.
So, I edited a little bit of it, and it works without any problems now. I don't know how to PR separately, and I committed it as "Fix comment bug".

Thanks.

**Comment by Ben10do** [comment]
Firstly, thanks for contributing – a trie does sound like a sensible data structure to use here! I'll take a look at the code for that a bit later 🙂

As for your second commit, you seem to have run into an issue reported as both #326 and #289. A fix for this would be appreciated, but as it's fixing an unrelated issue, please could you open a separate pull request for this fix? (It's generally preferred to have PRs that focus on solving one thing at a time!)

**Comment by jidoc01** [comment]
Yeah. I've searched how to make PRs separately, and knew that it just needs to create a temporary branch for each PL. I pushed another PL, so check it later please.

Thanks.

**Comment by AntonioND** [comment]
Add information to the commit message.

**Comment by jidoc01** [comment]
I understood that it's just okay to include keywords in the commit message, but it isn't.
I edited it. Thanks.

**Comment by AntonioND** [comment]
Commits need to be useful to track changes in the history of the code. If you're bored I advice you to read this: https://chris.beams.io/posts/git-commit/

File: include/asm/charmap.h [hunk 5]
```diff
@@ -13,14 +13,25 @@
 
 #define MAXCHARMAPS	512
 #define CHARMAPLENGTH	16
+#define MAXCHARNODES	MAXCHARMAPS * CHARMAPLENGTH + 1
```
Comment (AntonioND) [inline include/asm/charmap.h]: Add parentheses to this expression.

**Comment by jidoc01** [comment]
Oh, I've searched that post after I saw your comment before. Thanks to you, I could learn how to edit commit messages on vim. I am new to this git thing :D. Sorry to bother you.

And, I added parentheses to it.

Thanks.
